fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/cursors/ConcatCursorTest.java

```diff
@@ -0,0 +1,438 @@
+
+        //Row limit tests ...
+        ScanProperties scanProperties;
+        Integer maxRows = 51;
+        for (rowLimit = 0; rowLimit < maxRows; rowLimit++) {
+            ep = ExecuteProperties.newBuilder().setReturnedRowLimit(rowLimit).build();
+            scanProperties = new ScanProperties(ep);
+            if (rowLimit % 2 == 0) {
+                scanProperties.setReverse(true); //toggle between forward and reverse scans
+            } else {
+                scanProperties.setReverse(false);
+            }
+            concatRowLimitTest(scanProperties);
+        }
+
+        //Time limited tests
+        long maxTimeLimit = 100000;
+        long timeLimit;
+        for (timeLimit = 100; timeLimit < maxTimeLimit; timeLimit = timeLimit * 10) {
+            ep = ExecuteProperties.newBuilder().setTimeLimit(timeLimit).build();
+            scanProperties = new ScanProperties(ep);
+            concatTimeLimitTest(scanProperties);
+        }
+    }
+
+    @Test
+    public void sizeStatsCollectorTest() {
+        setupBaseData();
+        Subspace indexesSubspace = subspace.subspace(Tuple.from(FDBRecordStoreKeyspace.INDEX_SECONDARY_SPACE.key()));
+        SizeStatisticsCollector sc = SizeStatisticsCollector.ofSubspace(indexesSubspace);
+        long timeLimit = 1000;
+        ExecuteProperties ep = ExecuteProperties.newBuilder().setTimeLimit(timeLimit).build();
+
+        //create a list to append the totalSizeFuture too
+        List<Long> longs = Arrays.asList(1L, 2L, 3L);
+        ScanProperties scanProperties = ScanProperties.FORWARD_SCAN;
+        TriFunction<FDBRecordContext, ScanProperties, byte[], RecordCursor<Long>> f1 = (c, s, b) -> {
+            List<Long> l = (longs.stream().collect(Collectors.toList()));
+            if (s.isReverse()) {
+                Collections.reverse(l);
+            }
+            return RecordCursor.fromList(l, b);
+        };
+
+        // start the stats collection for the current index subspace in this fdb database
+        // create a lazy map cursor that attaches a callback for accessing the stats structure
+        CompletableFuture<Long> totalSizeFuture = fdb.runAsync(c -> sc.collectAsync(c, ep)).thenApply(b -> b ? sc.getTotalSize() : 0L);
+
+        TriFunction<FDBRecordContext, ScanProperties, byte[], RecordCursor<Long>> f2 = (c, s, b) -> {
+            return RecordCursor.fromFuture(totalSizeFuture);
+        };
+
+        //concat the list with the total size of the indexes and increase all the results an ordered of magnitude
+        ConcatCursor<Long> concatCursor = new ConcatCursor<>(context, scanProperties, f1, f2, null);
+        MapCursor<Long, Long> mapCursor = new MapCursor<>(concatCursor, l -> l * 10);
+        List<Long> result = mapCursor.asList().join();
+        Assertions.assertEquals(result.size(), longs.size() + 1);
+        Assertions.assertNotEquals(result.get(3), 0L);
+    }
+
+    public void concatListCursorTests(ScanProperties scanProperties) {
```

Comment:
I think you could genericize most of the logic here into a helper method that took a list of numbers and a number of chunks to split it into, like:

    // Break a list of 100 values out into 10 concatenated cursors
    RecordCursor<Integer> finalCursor = createConcatCursors(IntStream.range(1, 100).collect(Collectors.toList()), 10, isReversed);

Reply:
I get the gist. But the cursors "plans" being built in the test cases are actually trees and so building a generating function that changed the bushiness dimension would be more work. Given that this work is forming the basis for the stats work needed by the forecasting work would it make sense to open a separate Radar to generalize this? Thanks.